### PR TITLE
[MRG] Improve doc about caching methods

### PR DESCRIPTION
**Pull request by lesteve**
Recommend to cache pure functions that you call inside a method.

Add more information on caveats.

Initiated by #431.

### CODE REVIEW

**Comment by lesteve** [comment]
@smidm do let me know if something is not crystal clear with the new doc, in particular if this would not have helped you on your particular issue.

**Comment by smidm** [comment]
This would have pointed me in the right direction. Thanks.

**Comment by lesteve** [comment]
@aabadie @GaelVaroquaux maybe a quick look at this one?

File: doc/memory.rst [hunk 1]
```diff
@@ -324,26 +324,65 @@ Gotchas
...
+
 
     class Foo(object):
+        def __init__(self, args):
+            self.data = None
 
-        @mem.cache  # WRONG
-        def method(self, args):
-	    pass
+        def compute(self):
```
Comment (GaelVaroquaux) [inline doc/memory.rst]: I actually now prefer to use joblib not as an decorator, but as an explicit function transformation, and I would remove the decorator on compute_func and write the following:

self.data = mem.cache(compute_func)(self.arg1, self.arg2, 40)

Comment (lesteve) [inline doc/memory.rst]: I'll merge this PR and get rid of the mem.cache decorators in a separate PR.

Out of interest why do you do it this way to make it clearer that the function is cached where it is called, to keep the uncached function easily accessible, for some other reasons?

**Comment by GaelVaroquaux** [comment]
> to keep the uncached function easily accessible, for some other reasons?

For that specific reason.

**Pull request merged**